### PR TITLE
Infra/remove error blocks in ricos getContent

### PR DESCRIPTION
**Pull request by talevy17**

### CODE REVIEW

File: packages/ricos-editor/web/src/RicosEditor.tsx [hunk 1]
```diff
@@ -9,6 +9,7 @@ import './styles.css';
 import { RicosEditorProps, EditorDataInstance, RichContentChild } from './index';
 import { hasActiveUploads } from './utils/hasActiveUploads';
 import { convertToRaw } from 'wix-rich-content-editor/dist/lib/editorStateConversion';
+import errorBlocksRemover from './errorBlocksRemover';
```
Comment (Barackos) [inline packages/ricos-editor/web/src/RicosEditor.tsx:12]: Let's keep `RicosEditor` clean from additional imports. Idea: add a boolean argument to `getContentState`, let's say "shouldRemoveErrors" / something equivalent, and reposition the `errorBlocksRemover` to be right before the return statement.

**Review by Barackos — APPROVED** [review]

File: packages/ricos-common/web/src/RicosTypes.ts [hunk 2]
```diff
@@ -98,7 +98,7 @@ export interface ToolbarSettings {
 }
```
Comment (Barackos) [inline packages/ricos-common/web/src/RicosTypes.ts:99]:
```suggestion
interface ContentStateGetterArgs {
  shouldRemoveErrorBlocks?: boolean;
}
```

File: packages/ricos-editor/web/src/utils/editorUtils.ts [hunk 3]
```diff
@@ -32,7 +33,7 @@ export function createDataConverter(onContentChange?: OnContentChangeFunction):
     });
   };
 
-  const getContentState = () => {
+  const getContentState = ({ shouldRemoveErrorBlocks = false } = {}) => {
```
Comment (Barackos) [inline packages/ricos-editor/web/src/utils/editorUtils.ts:36]:
```suggestion
  const getContentState: ContentStateGetter = ({ shouldRemoveErrorBlocks } = {}) => {
```

**Review by Barackos — APPROVED** [review]
We can enforce object fields. Check my suggestions

File: examples/storybook/stories/UploadMedia/Utils/ErrorBlocksRemoalApp.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
+import React, { Component } from 'react';
```
Comment (shaulgo) [inline examples/storybook/stories/UploadMedia/Utils/ErrorBlocksRemoalApp.js:1]: fix file name removal

File: packages/ricos-editor/web/src/utils/editorUtils.ts [hunk 5]
```diff
@@ -32,7 +38,7 @@ export function createDataConverter(onContentChange?: OnContentChangeFunction):
     });
   };
 
-  const getContentState = () => {
+  const getContentState: ContentStateGetter = ({ shouldRemoveErrorBlocks = false } = {}) => {
```
Comment (shaulgo) [inline packages/ricos-editor/web/src/utils/editorUtils.ts:41]:
```suggestion
  const getContentState: ContentStateGetter = ({ shouldRemoveErrorBlocks = true } = {}) => {
```

File: packages/ricos-editor/web/src/RicosEditor.tsx [hunk 6]
```diff
@@ -68,12 +67,12 @@ export class RicosEditor extends Component<RicosEditorProps, State> {
 
   getToolbars = () => this.editor.getToolbars();
 
-  getContent = (postId?: string, forPublish?: boolean) => {
+  getContent = (postId?: string, forPublish?: boolean, shouldRemoveErrorBlocks = false) => {
```
[the remark this answers is not in the document]
Comment (shaulgo) [reply packages/ricos-editor/web/src/RicosEditor.tsx:72]: @Barackos what are postId and forPublish?

File: packages/ricos-editor/web/src/utils/errorBlocksRemover.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+      shouldKeepBlock = false;
+    } else if (entityType === galleryType) {
+      const items = data.items.filter(item => !item.error);
+      newEntityMap[key].data.items = items;
+    }
+  });
+  return { newEntityMap, shouldKeepBlock };
+};
+
+export default contentState => {
```
Comment (shaulgo) [inline packages/ricos-editor/web/src/utils/errorBlocksRemover.js:27]: let's give the function a name. Easier to read and debug.

File: packages/ricos-editor/web/src/utils/errorBlocksRemover.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  let { entityMap } = contentState;
+  const newBlocks = contentState.blocks.filter(block => {
+    const { entityRanges = [], type } = block;
+    const isAtomic = type === 'atomic';
+    if (isAtomic) {
+      const { newEntityMap, shouldKeepBlock } = modifyEntityMap(entityRanges, entityMap);
```
Comment (shaulgo) [inline packages/ricos-editor/web/src/utils/errorBlocksRemover.js:33]:
```suggestion
      const { newEntityMap, shouldKeepBlock } = modifyEntityMap(entityKey, entityMap);
```

File: packages/ricos-editor/web/src/utils/errorBlocksRemover.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+      delete newEntityMap[key];
+      shouldKeepBlock = false;
+    } else if (entityType === galleryType) {
+      const items = data.items.filter(item => !item.error);
+      newEntityMap[key].data.items = items;
+    }
+  });
+  return { newEntityMap, shouldKeepBlock };
+};
+
```
Comment (shaulgo) [inline packages/ricos-editor/web/src/utils/errorBlocksRemover.js:26]: isErroredBlock = (block, entityMap) => if (atomic && isErroredEntity(entityKey)

**Review by shaulgo — CHANGES_REQUESTED** [review]

File: packages/ricos-editor/web/src/RicosEditor.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (Barackos) [reply packages/ricos-editor/web/src/RicosEditor.tsx]: `postId` is a field we store in our BI when a publish occurs.
`forPublish` explicitly tells us that `getContent` is now being used to retrieve the final content for publish. We ask for this boolean so we won't trigger publish BI on `getContent()` calls that are unrelated to publish.

**Comment by shaulgo** [comment]
do you know how many consumers are using these functions? I think we
could improve the signature of this function. But it would be a breaking
change.

**Comment by Barackos** [comment]
@shaulgo From a quick search it seems that [Forum](https://github.com/wix-private/app-market/blob/cd50b1a971d56305b2b4038651e320d62fc80633/communities/communities-forum-client/src/app/components/rich-content-editor/rich-content-editor.js#L590) uses it, but I also remember Challenges (though I can't find their usage in the github search).

I'd love to improve the API. let's discuss.

File: packages/ricos-editor/web/src/utils/errorBlocksRemover.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+): boolean => type === 'atomic' && entityMap[entityKey].data?.error;
+
+export default function errorBlocksRemover(contentState: RicosContent): RicosContent {
+  const { entityMap } = contentState;
+  const newBlocks: RicosContentBlock[] = contentState.blocks.filter(
+    (block: RicosContentBlock): boolean => {
+      const { entityRanges = [], type } = block;
+      return !isErroredBlock(entityMap, entityRanges[0]?.key, type);
+    }
+  );
```
Comment (shaulgo) [inline packages/ricos-editor/web/src/utils/errorBlocksRemover.ts:33]:
```suggestion
  const newBlocks: RicosContentBlock[] = contentState.blocks.filter(!isErrorBlock);
    }
  );
```

File: packages/ricos-editor/web/src/utils/errorBlocksRemover.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+  return {
+    ...contentState,
+    blocks: newBlocks,
+    entityMap: modifyEntityMap(entityMap),
```
Comment (shaulgo) [inline packages/ricos-editor/web/src/utils/errorBlocksRemover.ts:37]:
```suggestion
    entityMap: removeEntitiesWithError(entityMap),
```

**Review by shaulgo — APPROVED** [review]
add test for the removal

**Pull request merged**